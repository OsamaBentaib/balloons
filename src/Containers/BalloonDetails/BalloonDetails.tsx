import React from "react";
import { useSelector } from "react-redux";
import { useQuery } from "urql";
import BalloonDetailsContainer from "../BalloonDetailsContainer";

const BalloonQuery = `
query balloon($id: String!){
  balloon(id:$id){
    id
    imageUrl
    name
    price
    variant
    color
    description
    availableSince
  }
}
`;
export default function BalloonDetails() {
  // get the state
  const initialState: InitialState | undefined = useSelector(
    (initialState: InitialState) => initialState
  );

  // the the states that we need
  const {
    balloonDetails: { balloonId },
  } = initialState;

  // fetch the query
  const [result] = useQuery({
    query: BalloonQuery,
    variables: { id: balloonId },
  });

  const { data, fetching, error } = result;
  return (
    <BalloonDetailsContainer data={data} fetching={fetching} error={error} />
  );
}
