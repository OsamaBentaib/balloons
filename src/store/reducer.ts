import {
  ADD_REMOVE_CART,
  CLOSE_CART,
  CLOSE_DETAILS,
  OPEN_CART,
  OPEN_DETAILS,
  RESET_FILTERS,
  SET_FILTER_WITH_COLOR,
  SET_FILTER_WITH_VARIANT,
  SET_PAGINATION,
  SET_SORT_BY,
} from "./actionTypes";
import { sortInput } from "./paramsTypes";

/**
 * as default we need to check on localstorage if there's a cart 
 * and get the cart string and convert it to a JSON object
 */
const storedItems = localStorage.getItem("_cart");
let cartItems: CartItem[] = [];
// check for the type of storedItems
if (typeof storedItems === "string") {
  // convert the storedItems string to a json object
  cartItems = JSON.parse(storedItems);
}
const initialState: InitialState = {
  balloonDetails: {
    isDetailsOpen: false, // as default the balloon details won't be displayed
    balloonId: null, // the id should be null as default
  },
  cart: {
    isCartOpen: false, // as default the cart won't be displayed
    items: cartItems, // as default the cart would be empty or they fetch from storage
  },
  paramsQueryVariables: {
    filter: {}, // as default it's won't be any filter with balloons
    sort: sortInput.ID, // as dafault we will fetch
  },
};


const reducer = (
  state: InitialState = initialState,
  action: Action
): InitialState => {
  // set the right action to update the state
  switch (action.type) {
    case ADD_REMOVE_CART:
      // remove the previews cart and replace it
      localStorage.removeItem("_cart");
      const items = action.payload.cart.items;
      // save the json object as string inside the localstorage
      localStorage.setItem("_cart", JSON.stringify(items));
      return {
        ...state,
        cart: { ...state.cart, items: items },
      };
    case OPEN_CART:
      return {
        ...state,
        cart: { ...state.cart, isCartOpen: true },
      };
    case CLOSE_CART:
      return {
        ...state,
        cart: { ...state.cart, isCartOpen: false },
      };
    case OPEN_DETAILS:
      return {
        ...state,
        balloonDetails: action.payload.balloonDetails,
      };
    case CLOSE_DETAILS:
      return {
        ...state,
        balloonDetails: action.payload.balloonDetails,
      };
    case SET_SORT_BY:
      return {
        ...state,
        paramsQueryVariables: {
          ...state.paramsQueryVariables,
          sort: action.payload,
        },
      };
    case SET_FILTER_WITH_COLOR:
      return {
        ...state,
        paramsQueryVariables: {
          ...state.paramsQueryVariables,
          filter: {
            ...state.paramsQueryVariables.filter,
            color: action.payload,
          },
        },
      };
    case SET_FILTER_WITH_VARIANT:
      return {
        ...state,
        paramsQueryVariables: {
          ...state.paramsQueryVariables,
          filter: {
            ...state.paramsQueryVariables.filter,
            variant: action.payload,
          },
        },
      };
    case RESET_FILTERS:
      return {
        ...state,
        paramsQueryVariables: {
          filter: {}, // as default it's won't be any filter with balloons
          sort: sortInput.ID, // as dafault we will fetch
        },
      };
    case SET_PAGINATION:
      return {
        ...state,
        paramsQueryVariables: {
          ...state.paramsQueryVariables,
          before: action.payload.before,
          after: action.payload.after,
        },
      };

    default:
      return { ...state };
  }
};

export default reducer;
