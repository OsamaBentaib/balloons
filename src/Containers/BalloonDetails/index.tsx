import BalloonDetails from './BalloonDetails'
export default BalloonDetails;