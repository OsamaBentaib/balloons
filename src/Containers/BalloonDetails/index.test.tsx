import Enzyme, { mount, shallow } from "enzyme";
import { Provider as URQLProvider } from "urql";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import { never } from "wonka";
import { Provider } from "react-redux";
import { createStore } from "redux";
import reducer from "../../store/reducer";
import BalloonDetails from "./BalloonDetails";

const initialState: InitialState = {
  balloonDetails: {
    isDetailsOpen: false, // as default the balloon details won't be displayed
    balloonId: null, // the id should be null as default
  },
  cart: {
    isCartOpen: false, // as default the cart won't be displayed
    items: [], // as default the cart would be empty or they fetch from storage
  },
  paramsQueryVariables: {
    filter: {}, // as default it's won't be any filter with balloons
    sort: "ID", // as dafault we will fetch
  },
};

Enzyme.configure({ adapter: new Adapter() });

describe("<BalloonDetails /> renders unit test", () => {
  const mockClient = {
    executeQuery: jest.fn(() => never),
    executeMutation: jest.fn(() => never),
    executeSubscription: jest.fn(() => never),
  };
  const mockStore = createStore(reducer, initialState);
  const getWrapper = () =>
    mount(
      //@ts-ignore
      <URQLProvider value={mockClient}>
        <Provider store={mockStore}>
          <BalloonDetails />
        </Provider>
      </URQLProvider>
    );
  it("call the query once when it's renders", () => {
    getWrapper();
    expect(mockClient.executeQuery).toBeCalledTimes(1);
  });
  it("matches snapshot", () => {
    const responseState = {
      executeQuery: () =>
        fromValue({
          data: {
            balloon: {
              __typename: "Balloon",
              id: "blue",
              imageUrl: "/balloons/heart-blue.jpg",
              name: "Light Blue Heart Balloon",
              price: 10,
              variant: "HEART",
              color: "BLUE",
              description:
                "A blue balloon, great for occasions involving hearts.",
              availableSince: "2020-05-13T13:00:00Z",
            },
          },
        }),
    };
    const wrapper = shallow(
      //@ts-ignore
      <URQLProvider value={responseState}>
        <Provider store={mockStore}>
          <BalloonDetails />
        </Provider>
      </URQLProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
function fromValue(arg0: {
  data: {
    balloon: {
      __typename: string;
      id: string;
      imageUrl: string;
      name: string;
      price: number;
      variant: string;
      color: string;
      description: string;
      availableSince: string;
    };
  };
}) {
  throw new Error("Function not implemented.");
}
