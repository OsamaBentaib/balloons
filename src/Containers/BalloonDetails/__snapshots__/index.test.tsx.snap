// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BalloonDetails /> renders unit test matches snapshot 1`] = `ShallowWrapper {}`;
